### DISCUSSION TITLE
What is the origin of the platlib/_preload_exname_extver.pth file?

### DISCUSSION THREAD

**Discussion by ndw** [pypa/cibuildwheel#1440] [General]
In January, I used a `cibuildwheel` [task on GitHub](https://github.com/Saxonica/BuildWheels/) to build wheels for an external C library.  A few days ago, I used it again to build wheels for an updated version of the library.

The version built a few days ago includes an additional file in the wheel:

```
saxoncpe-12.1.0.data/platlib/_preload_saxoncpe_12.1.0.pth
```

That file isn't present in the "12.0.0" version of the wheel and it appears to cause problems with virtual environments on Windows:

```
Error processing line 1 of C:\Users\Norman Walsh\saxon-test-1\lib\site-packages\_preload_saxonche_12.1.0.pth:

  Traceback (most recent call last):
    File "C:\Python39\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<string>", line 18, in <module>
  FileNotFoundError: .load-order-saxonche-12.1.0 not found

Remainder of file ignored
SaxonC-HE 12.1 from Saxonica
```

I can't work out if this is caused by something that changed in the configuration of our build, something that changed in `cibuildwheel` or something else.

Hints, suggestions, workarounds, etc. most humbly solicited.

**Comment by joerick** [comment]
Your build script still has v2.11.1 pinned by the looks of it, so I'm guessing the change will be in the code you're building. I can't seem to access that, is it accessible somewhere?

**Comment by ndw** [comment]
We're not building from exclusively the open-source editions of the product, so I don't have all the sources in an easily accessible place, but here's a link to what the `cibuildwheel` tool builds from: https://dev.saxonica.com/_wheels/

(The github action downloads and unpacks that before it begins.)

I don't think we made any changes to the configuration, but I've been wrong before.

**Comment by henryiii** [comment]
That's an editable install artifact, I'm going to bet. Some backends use generated pth files.  Is there an editable install happening somewhere, like for coverage or something? What build backend are you using?

**Reply by ndw** [reply]
We combine a bunch of Java and some C/C++ with GraalVM and construct DLL (or platform-appropriate dynamic library) and some Python wheel configuration. That gets handed to `cibuildwheel` and we upload the resulting wheels to PyPi. We don't do anything to them after the `windows.wheel.tar.gz` file you can find in the link above. For the 12.0 release, no `.pth` files, for the 12.1 release, `.pth` files. We didn't change anything about the build configuration (on purpose), so I'm at a loss.

(I'm also, obviously, skating right along the very edge of what I know about building Python wheels!)

**Reply by henryiii** [reply]
I don't specifically recognize that `.pth` file prefix - usually they have `editable` somewhere in the name. This is the one I'm aware of from setuptools, which would be `__editable__.saxoncpe.pth`, I think, also I think added in 62 or so, which you seem to be pinning an older version: https://github.com/pypa/setuptools/blob/be6c0218bcba78dbd4ea0b5a8bb9acd5d5306240/setuptools/command/editable_wheel.py#L389

**Reply by ndw** [reply]
I can certainly try to pin to more recent versions of `setuptools` and `cython` and see what happens. I don't recall precisely how we arrived at those numbers (though I have a vague recollection that it was intentional).

**Reply by henryiii** [reply]
Setuptools is easy - NumPy recommended capping to <60. It was a horrible recommendation, basically due to internal details and NumPy not planning on supporting setuptools anymore (and can currently be fixed with an envvar setting instead of a cap!), but that caused a ton of libraries to add caps, even if it was completely irrelevant to them.

Though I don't expect either of those to add the file - but don't see anything else that would add it - setuptools looks like it's the only thing running that would have the necessary info to add such a file with the name and version in it. (well, pip too). FYI, setuptools has native Cython support, you don't need `cythonize`.

Maybe try setting the build frontend to build instead of pip?

**Reply by ndw** [reply]
Can you point me to some documentation that explains what "setting the build frontend to build instead of pip" means? I am a little out of my depth here and that's probably not helping.

**Reply by henryiii** [reply]
Well, not important now, but https://cibuildwheel.readthedocs.io/en/stable/options/#build-frontend was what I was referring to.

**Reply by ndw** [reply]
Updating setuptools had no effect, but that's not a surprise given the dicoveries about delvewheel below.

**Comment by joerick** [comment]
I did [a little sleuthing on that filename](https://sourcegraph.com/search?q=context:global+content:_preload.*%5C.pth+lang:Python+&patternType=regexp&sm=1&groupBy=repo) and it looks like it's a feature of delvewheel...

https://github.com/adang1345/delvewheel/blob/master/delvewheel/_wheel_repair.py#L816-L835

**Reply by ndw** [reply]
Nice sleuthing!

```
2023-03-20T15:26:05.4254532Z ##[group]Repairing wheel...
2023-03-20T15:26:05.4254870Z 
2023-03-20T15:26:05.4257123Z + delvewheel repair --add-path libs/win -w C:\Users\runneradmin\AppData\Local\Temp\cibw-run-fh7ny6cj\cp38-win_amd64\repaired_wheel C:\Users\runneradmin\AppData\Local\Temp\cibw-run-fh7ny6cj\cp38-win_amd64\built_wheel\saxoncee-12.1.0-cp38-cp38-win_amd64.whl
2023-03-20T15:26:10.4275582Z C:\Users\runneradmin\AppData\Local\Temp\cibw-run-fh7ny6cj\cp38-win_amd64\build\venv\lib\site-packages\delvewheel\_wheel_repair.py:819: UserWarning: 
2023-03-20T15:26:10.4276735Z repairing C:\Users\runneradmin\AppData\Local\Temp\cibw-run-fh7ny6cj\cp38-win_amd64\built_wheel\saxoncee-12.1.0-cp38-cp38-win_amd64.whl
2023-03-20T15:26:10.4277548Z The wheel contains a top-level extension module and has a delay-load
2023-03-20T15:26:10.4277919Z finding DLL dependencies
2023-03-20T15:26:10.4278442Z dependency, so we are injecting some code that runs at Python startup
2023-03-20T15:26:10.4279009Z copying DLLs into saxoncee-12.1.0.data\platlib
2023-03-20T15:26:10.4279514Z to preload the dependencies. End users who have Python processes that
2023-03-20T15:26:10.4279856Z mangling DLL names
2023-03-20T15:26:10.4280306Z were running before the wheel was installed should start new Python
2023-03-20T15:26:10.4280609Z calculating DLL load order
2023-03-20T15:26:10.4281055Z instances after installing the wheel.
2023-03-20T15:26:10.4281397Z updating saxoncee-12.1.0.dist-info\RECORD
2023-03-20T15:26:10.4281835Z   warnings.warn(textwrap.fill(
2023-03-20T15:26:10.4282134Z repackaging wheel
2023-03-20T15:26:10.4282752Z fixed wheel written to C:\Users\runneradmin\AppData\Local\Temp\cibw-run-fh7ny6cj\cp38-win_amd64\repaired_wheel\saxoncee-12.1.0-cp38-cp38-win_amd64.whl
2023-03-20T15:26:10.7709588Z ##[endgroup]
```

Now what the heck is a `delay-load`, why was it not true in 12.0, and how do I make it go away :-(

**Reply by henryiii** [reply]
Fantastic! Totally forgot about delvewheel running on this. :)

**Reply by henryiii** [reply]
Well, as an emergency workaround, you could pin an older version, since it was working before. Might be worth opening an issue over on delvewheel.

**Reply by henryiii** [reply]
The info on delvewheel is at https://github.com/adang1345/delvewheel. I see it discusses delay load a bit, but not how to disable it AFAICT.

**Comment by ndw** [comment]
I had a brief hope that maybe using a different version of GraalVM (22.3.1 instead of 22.3.0) was relevant, but no :-(

I don't see where I specify an particular version of delvewheel, is that under my control?

**Reply by henryiii** [reply]
Yes, you are pip installing it, just add `==<some_version>`.

Edit: looks like you found that https://github.com/Saxonica/BuildWheels/blob/614c922637da20c04cdc578704a7a5947802f526/.github/workflows/wheels.yml#L43 :)

**Reply by ndw** [reply]
Yes. Thank you. *blush* Trying 1.2.0 now which I think was the current version last time our build succeeded.

**Comment by ndw** [comment]
Right. Going back to delvewheel 1.2.0 fixes the problem in as much as there are no `.pth` files. Whether there are other possible consequences, I'm not sure.

Thank you, everyone, for your help! I really appreciate it!
